clients/venice-client/src/main/java/com/linkedin/venice/fastclient/transport/GrpcTransportClient.java

```diff
@@ -22,209 +24,278 @@
+    if (sslFactory == null) {
+      return InsecureChannelCredentials.create();
+    }
+
     try {
       TlsChannelCredentials.Builder tlsBuilder = TlsChannelCredentials.newBuilder()
           .keyManager(GrpcSslUtils.getKeyManagers(sslFactory))
           .trustManager(GrpcSslUtils.getTrustManagers(sslFactory));
-      channelCredentials = tlsBuilder.build();
+      return tlsBuilder.build();
     } catch (Exception e) {
       throw new VeniceClientException(
           "Failed to initialize SSL channel credentials for Venice gRPC Transport Client",
           e);
     }
   }
 
-  protected ManagedChannel getChannel(String serverAddress) {
-    if (!nettyAddressToGrpcAddressMap.containsKey(serverAddress)) {
-      throw new VeniceException("No grpc server found for address: " + serverAddress);
+  @VisibleForTesting
+  VeniceClientRequest buildVeniceClientRequest(String[] requestParts, byte[] requestBody, boolean isSingleGet) {
+    VeniceClientRequest.Builder requestBuilder = VeniceClientRequest.newBuilder()
+        .setResourceName(requestParts[4])
+        .setIsBatchRequest(!isSingleGet)
+        .setMethod(isSingleGet ? HttpMethod.GET.name() : HttpMethod.POST.name());
+
+    if (isSingleGet) {
+      requestBuilder.setKeyString(requestParts[6]);
+      requestBuilder.setPartition(Integer.parseInt(requestParts[5]));
+    } else {
+      requestBuilder.setKeyBytes(ByteString.copyFrom(requestBody));
     }
 
-    return serverGrpcChannels.computeIfAbsent(
-        serverAddress,
-        k -> Grpc.newChannelBuilder(nettyAddressToGrpcAddressMap.get(k), channelCredentials).build());
+    return requestBuilder.build();
   }
 
-  protected VeniceReadServiceGrpc.VeniceReadServiceStub getStub(ManagedChannel channel) {
+  @VisibleForTesting
+  VeniceReadServiceGrpc.VeniceReadServiceStub getOrCreateStub(String serverAddress) {
+    String grpcAddress = getGrpcAddressFromServerAddress(serverAddress);
+
+    ManagedChannel channel = serverGrpcChannels
+        .computeIfAbsent(serverAddress, k -> Grpc.newChannelBuilder(grpcAddress, channelCredentials).build());
+
     return stubCache.computeIfAbsent(channel, VeniceReadServiceGrpc::newStub);
   }
 
-  @Override
-  public CompletableFuture<TransportClientResponse> get(String requestPath, Map<String, String> headers) {
-    return handleRequest(requestPath, headers, null, true);
+  /**
+   * Maps the given server address to GRPC server address. Typically, the mapped server address only differs in the
+   * port. Fetch the GRPC server address for a given server address if available in the #nettyServerToGrpcAddress
+   * map if not, strip the port from the original server address and replace it with the grpc port.
+   * <b>Note:</b> We have the GRPC port hard coded and operate with static port assignment for GRPC
+   */
+  @VisibleForTesting
+  String getGrpcAddressFromServerAddress(String serverAddress) {
```

Comment:
I didn't realize how static the mapping needed to be. This means if we'd like to configure the port to something different than the fallback `GRPC_PORT` we need to provide a config map to the client that contains ALL the servers that's potentially reachable by that client (could be multiple clusters). That doesn't sound very appealing... if that's the case shall we also make the fallback configurable in case we want to change the server side hardcode to something that's not `23900`.

Reply:
Agreed. I thought about it and was going to put #1 in this PR given its short term. While I was going do option 2 or 3 in the mid/long term if things look good as part of taking this to production.

My only reason for abandoning option 1 was introducing a new client configuration and the debt that tags along in terms of deprecating it if we are going to jump to option 2 or 3. 

How easy with venice when it comes to deprecating configurations? Would piggy backing the existing `nettyServerToGrpcAddress` be flexible enough at the same time prevent introducing public facing configurations as part of client?

Reply:
Client config for GRPC port doesn't sound great even for the first customer as we are leaking backend config into client.
One workaround is to maintain a global GRPC port config, which will be managed by us and if we want to change it, we will change the global config instead of asking everyone update it, and this can work with your current code change.
Global config for GRPC port is not great as option 2/3 as the customer still needs to do a new deployment to pick up the new port, but it might be acceptable as we hide it from the customer and still manage it in a centralized fashion for the time being.

Reply:
I was mostly referring to introducing a client config within `ClientConfig` be it internal. I think I am on the same page is making `23900` configurable through some means. Do you guys think we can leverage the existing map that we already use to pass this port to the `GrpcTransportClient`? If so, we can get away from introducing another configuration within venice repo but still populate the map externally using a global configuration keeping the internal contract more flexible for deprecation.

Reply:
Shall we just have a single config for the GPRC port instead of a map?
The assumption is that all the instances will use the same port in the same cluster.
And this would only be used by the benchmarking, not the prod use case, right?